Add basic support for Websocket connection retry

markadr opened the pull request:
Found that if the WebSocket service side of this was not already started or had to be restarted during a run, the WebSocket client wouldn't recover so no DMD updates would be sent.

This change is one approach at allowing the connection to recover if the service is not already running when a game starts or gets restarted during the lifetime of a game run.

Tested using both the dmdext cli tool and via VPX game setup that uses the network streaming option.

Retry feature is disabled by default and when enabled, a retry interval of 5 seconds. The retry interval can be changed via another option to be more or less frequent, lowest interval is 1 second.

freezy commented on Console/Common/BaseOptions.cs:
```diff
@@ -67,6 +67,10 @@ internal abstract class BaseOptions : IConfiguration
 
+		[Option("retry", HelpText = "If set, retry making Websocket connections. Default: false" )]
+		public bool WebsocketRetry { get; set; } = false;
+		[Option("retryInterval", HelpText = "Delay between Websocket connection retry attempts, in seconds. Default: 5 seconds")]
```
Would prefer `retry-interval` to stay coherent with the other multi-word options.

markadr replied:
Makes sense, will update

freezy commented on LibDmd/Output/Network/NetworkStream.cs:
```diff
@@ -12,41 +14,59 @@ public class NetworkStream : IGray2Destination, IGray4Destination, IColoredGray2
 		private Color _color = RenderGraph.DefaultColor;
 		private Color[] _palette;
 
+		private TimeSpan PollForWSReconnection { get; set; } = TimeSpan.FromSeconds(10);
```
Doesn't seem to be used?

markadr replied:
From an earlier iteration, will remove.

freezy commented on LibDmd/Output/Network/NetworkStream.cs:
```diff
@@ -69,6 +89,20 @@ private void OnError(object sender, ErrorEventArgs e)
+			if (!_disposed && _retry)
+			{
+				// Create a one-shot timer set to go off after the _retryInterval has passed.
+				Timer retryTimer = new Timer(_retryInterval);
+				retryTimer.Elapsed += OnReconnect;
+				retryTimer.AutoReset = false;
+				retryTimer.Enabled = true;
```
I *think* even if it's a one-time shot, your timer will leak, i.e. won't be disposed when `OnReconnect` is executed. So I think you'll need to manually dispose it.

So i'm reading this right that you're actually not polling for retrial, but only triggering once? Does `_client.Connect()` re-trigger an `OnClose()` when it fails connecting?

markadr replied:
Very limited .Net/C# dev experience, my expectation was that the timer would be done once it fired the event but easy enough to also call close on it in the event method.

> So i'm reading this right that you're actually not polling for retrial, but only triggering once? 
> Does _client.Connect() re-trigger an OnClose() when it fails connecting?

From my local testing/experiments, the OnClose() method is called when a connection cannot be made or becomes disconnected after a connection has been established. OnClose() is called again should the connection break and this continues for the life of the app, no polling required.

I started out completely recreating the Websocket object on each re-connection attempt but found I could reuse the existing one and just call Connect() again. Haven't dug into the docs about this so not sure if the API was meant to allow this.

freezy replied:
Sounds good!

markadr commented:
> 
> 
> Awesome, thanks a lot!
> 
> Just a few minor things to change. Out of curiosity, what are you using the websocket driver for?

Using this version locally with my Android app (mentioned in the forum a few weeks ago) that can act as a DMD screen, saves me having to have a 3rd screen/use backglass on my limited PC setup.

markadr commented:
Branch updated based on PR feedback:
- Updated retryInterval to be retry-interval
- Moved to reusing retry timer as docs point out that a one-shot timer can be used again by calling Start() method.
- Added the new retry options with comments into the DmdDevice.ini file.
- Reworded help tips on new retry items.

freezy commented:
Cheers!

Will test later this week-end and merge if everything's okay.

freezy commented:
> Using this version locally with my Android app (mentioned in the forum a few weeks ago) that can act as a DMD screen, saves me having to have a 3rd screen/use backglass on my limited PC setup.

Yes, I remember. Wanted to check it out, but didn't have the time yet. Sounds really cool, will add it to the README once tested. :)

[Reference](https://github.com/markadr/dmd-viewer)

Thanks again for the PR!
